learn/json_preprocessor.go

```diff
@@ -0,0 +1,51 @@
+package learn
+
+import (
+	"io"
+	"unicode"
+)
+
+type stripControlCharactersReader struct {
+	wrapped io.Reader
+}
+
+func newStripControlCharactersReader(wrapped io.Reader) stripControlCharactersReader {
+	return stripControlCharactersReader{wrapped: wrapped}
+}
+
+// Read up to len(p) bytes, removing any unescaped control characters found.
+// Removed characters do not count toward the total bytes read.  Returns the
+// number of bytes read.
+func (r stripControlCharactersReader) Read(p []byte) (n int, err error) {
+	pIdx := 0
+	buf := make([]byte, len(p))
+
+	// Read up to len(p) bytes, then discard any control characters.  Continue
+	// reading (and discarding control characters) until p is full or there are
+	// no more bytes to read.
+	for pIdx < len(p) {
+		remaining := len(p) - pIdx
+		bufSlice := buf[:remaining]
+
+		var bufN int
+		bufN, err = r.wrapped.Read(bufSlice)
+
+		// Copy from buf to p, skipping unescaped control characters.
+		for _, r := range string(bufSlice) {
+			if unicode.IsControl(r) {
+				continue
+			}
```

Comment:
Seems like we could do without the `[]byte`-to-`string` conversion (which is a copy!). Instead, we could iterate over the `[]byte` and check whether `r <= 0x1f`.

Reply:
Good call, fixed in 451cae0a07f2548db5b0267ff872ef900ef33c88.